strict-mvar/src/Control/Concurrent/Class/MonadMVar/Strict/Checked.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE BangPatterns  #-}
+{-# LANGUAGE CPP           #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeFamilies  #-}
+{-# LANGUAGE TypeOperators #-}
+
+-- | This module corresponds to 'Control.Concurrent.MVar' in "base" package
+--
+module Control.Concurrent.Class.MonadMVar.Strict.Checked
+  ( -- * StrictMVar
+    StrictMVar
+  , castStrictMVar
+  , toLazyMVar
+  , fromLazyMVar
+  , newEmptyMVar
+  , newEmptyMVarWithInvariant
+  , newMVar
+  , newMVarWithInvariant
+  , takeMVar
+  , putMVar
+  , readMVar
+  , swapMVar
+  , tryTakeMVar
+  , tryPutMVar
+  , isEmptyMVar
+  , withMVar
+  , withMVarMasked
+  , modifyMVar_
+  , modifyMVar
+  , modifyMVarMasked_
+  , modifyMVarMasked
+  , tryReadMVar
+    -- * Re-exports
+  , MonadMVar
+  ) where
+
+import           Control.Monad.Class.MonadMVar (MonadMVar)
+import qualified Control.Monad.Class.MonadMVar as Lazy
+import           GHC.Stack (HasCallStack)
+
+--
+-- StrictMVar
+--
+
+type LazyMVar m = Lazy.MVar m
+
+-- | A strict MVar with invariant checking.
+--
+-- There is a weaker invariant for a 'StrictMVar' than for a 'StrictTVar' (see
+-- the @strict-stm@ package): although all functions that modify the
+-- 'StrictMVar' check the invariant, we do /not/ guarantee that the value inside
+-- the 'StrictMVar' always satisfies the invariant. Instead, we /do/ guarantee
+-- that if the 'StrictMVar' is updated with a value that does not satisfy the
+-- invariant, an exception is thrown. The reason for this weaker guarantee is
+-- that leaving an 'MVar' empty can lead to very hard to debug "blocked
+-- indefinitely" problems.
+data StrictMVar m a = StrictMVar {
+    -- | The invariant that is checked whenever the 'StrictMVar' is updated.
+    invariant :: !(a -> Maybe String)
+  , mvar      :: !(LazyMVar m a)
+  }
```

Comment:
With `Strict.TVar` we provide a single API for strict `TVar`s with and without invariants.  Switching from one to the other is then just a matter of changing a flag rather than importing a different module.  This might be nicer for using it just on CI.  I'd prefer if we follow the same approach here as well.

Reply:
@dcoutts what do you think?

Reply:
I think for a library, it's not great to have a build flag because it's global to all uses of the library. I think for a lib it'd be better to provide inter-changable modules. Applications can still use a build flag (with cpp) if they want to.

Reply:
These modules are not interchangeable, `Checked -> Unchecked` will cause type errors.  I don't mean we shouldn't do what @dcoutts suggests, rather add the necessary interfaces to the unchecked module (although if we do that some functions will be confusing without good haddocks).

Reply:
I will ensure that both modules are drop-in replacements for each other, and to make this clear in the documentation.

Reply:
The interfaces should now match, and I've updated documentation to explain why the unchecked module now includes `newMVarWithInvariant` for example. Let me know what you think